File: .agents/skills/tinyengine-dsl-generator/scripts/check_css.py
#!/usr/bin/env python3
"""
TinyEngine CSS Syntax Checker

检查DSL中的CSS字段是否有语法错误。

支持三种模式：
1. basic: 基础语法检查（默认，无需额外依赖）
2. tinycss2: 使用 tinycss2 库（需要安装：pip install tinycss2）
3. postcss: 使用 postcss 命令行工具（需要安装：npm install -g postcss）

用法：
  python3 check_css.py <dsl-file> [mode]
  python3 check_css.py <dsl-file> postcss  # 使用postcss检查
  python3 check_css.py <dsl-file> tinycss2 # 使用tinycss2检查
"""

import json
import re
import subprocess
import sys
from typing import Dict, List, Any


class CSSChecker:
    """CSS语法检查器基类"""

    def __init__(self, dsl_data: Dict[str, Any]):
        self.dsl = dsl_data
        self.errors = []
        self.warnings = []

    def check(self) -> bool:
        """检查CSS语法"""
        page_content = self.dsl.get('page_content', self.dsl)
        css_string = page_content.get('css', '')

        if not css_string:
            self.warnings.append("No CSS field found")
            return True

        return self._check_css(css_string)

    def _check_css(self, css: str) -> bool:
        """子类实现具体的检查逻辑"""
        raise NotImplementedError

    def report(self) -> str:
        """生成报告"""
        lines = []
        if self.errors:
            lines.append("❌ CSS Errors:")
            for error in self.errors:
                lines.append(f"  - {error}")
        if self.warnings:
            lines.append("⚠️ CSS Warnings:")
            for warning in self.warnings:
                lines.append(f"  - {warning}")
        if not self.errors and not self.warnings:
            lines.append("✅ CSS check passed!")
        return "\n".join(lines)


class BasicCSSChecker(CSSChecker):
    """基础CSS语法检查器（无需额外依赖）"""

    def _check_css(self, css: str) -> bool:
        """基础检查：括号匹配、基本语法"""
        # 检查括号匹配
        stack = []
        i = 0
        while i < len(css):
            char = css[i]
            if char in '{':
                stack.append((char, i))
            elif char in '}':
                if not stack or stack[-1][0] != '{':
                    self.errors.append(f"Unmatched '}}' at position {i}")
                    return False
                stack.pop()
            elif char in '(':
                stack.append((char, i))
            elif char in ')':
                if not stack or stack[-1][0] != '(':
                    self.errors.append(f"Unmatched ')' at position {i}")
                    return False
                stack.pop()
            i += 1

        if stack:
            for char, pos in stack:
                self.errors.append(f"Unclosed '{char}' at position {pos}")
            return False

        # 移除注释进行检查
        css_no_comments = re.sub(r'/\*.*?\*/', '', css, flags=re.DOTALL)

        # 检查是否有CSS规则
        if '{' not in css_no_comments:
            self.warnings.append("CSS may not contain any rules")

        # 检查分号使用
        rules = re.findall(r'\{([^}]*)\}', css_no_comments)
        for rule in rules:
            properties = rule.split(';')
            for prop in properties[:-1]:  # 最后一个可能为空
                prop = prop.strip()
                if prop and ':' not in prop:
                    self.warnings.append(f"Property without colon: {prop[:50]}")

        return len(self.errors) == 0


class TinyCSS2Checker(CSSChecker):
    """使用tinycss2库的CSS检查器"""

    def _check_css(self, css: str) -> bool:
        try:
            import tinycss2
        except ImportError:
            self.errors.append(
                "tinycss2 not installed. Install with: pip install tinycss2"
            )
            return False

        # 使用tinycss2解析CSS
        try:
            # 解析CSS规则列表
            rules = tinycss2.parse_stylesheet(css, skip_comments=False)

            # tinycss2会抛出解析错误
            for rule in rules:
                if rule.type == 'error':
                    self.errors.append(f"Parse error: {rule.message}")

            return len(self.errors) == 0

        except Exception as e:
            self.errors.append(f"Failed to parse CSS: {e}")
            return False


class PostCSSChecker(CSSChecker):
    """使用postcss命令行工具的CSS检查器"""

    def _check_css(self, css: str) -> bool:
        # 检查postcss是否可用
        try:
            result = subprocess.run(
                ['postcss', '--version'],
                capture_output=True,
                text=True,
                timeout=5
            )
            if result.returncode != 0:
                self.errors.append("postcss command failed. Install with: npm install -g postcss")
                return False
        except FileNotFoundError:
            self.errors.append("postcss not found. Install with: npm install -g postcss")
            return False
        except subprocess.TimeoutExpired:
            self.errors.append("postcss command timed out")
            return False

        # 将CSS写入临时文件
        import tempfile
        with tempfile.NamedTemporaryFile(mode='w', suffix='.css', delete=False) as f:
            f.write(css)
            temp_file = f.name

        try:
            # 使用postcss解析CSS
            result = subprocess.run(
                ['postcss', temp_file],
                capture_output=True,
                text=True,
                timeout=10
            )

            # postcss的错误输出
            if result.stderr:
                # 过滤掉警告（如<css input>:3:3: Yellow color）
                errors = []
                warnings = []
                for line in result.stderr.strip().split('\n'):
                    if line:
                        if any(w in line.lower() for w in ['warning', 'yellow']):
                            warnings.append(line)
                        else:
                            errors.append(line)

                self.errors.extend(errors)
                self.warnings.extend(warnings)

            return len(self.errors) == 0

        except subprocess.TimeoutExpired:
            self.errors.append("postcss command timed out")
            return False
        finally:
            import os
            try:
                os.unlink(temp_file)
            except:
                pass


def main():
    """主函数"""
    if len(sys.argv) < 2:
        print("Usage: check_css.py <dsl-file> [mode]")
        print("  mode: basic (default), tinycss2, postcss")
        sys.exit(1)

    file_path = sys.argv[1]
    mode = sys.argv[2] if len(sys.argv) > 2 else 'basic'

    # 读取DSL文件
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            dsl_data = json.load(f)
    except json.JSONDecodeError as e:
        print(f"❌ Invalid JSON: {e}")
        sys.exit(1)
    except FileNotFoundError:
        print(f"❌ File not found: {file_path}")
        sys.exit(1)

    # 选择检查器
    checkers = {
        'basic': BasicCSSChecker,
        'tinycss2': TinyCSS2Checker,
        'postcss': PostCSSChecker
    }

    checker_class = checkers.get(mode)
    if not checker_class:
        print(f"❌ Unknown mode: {mode}")
        print(f"Available modes: {', '.join(checkers.keys())}")
        sys.exit(1)

    checker = checker_class(dsl_data)
    is_valid = checker.check()
    print(checker.report())
    sys.exit(0 if is_valid else 1)


if __name__ == '__main__':
    main()

File: .agents/skills/tinyengine-dsl-generator/scripts/check_css.mjs
#!/usr/bin/env node
/**
 * TinyEngine CSS Syntax Checker
 *
 * 检查DSL中的CSS字段是否有语法错误（基础模式：括号匹配、基本语法，无需额外依赖）。
 *
 * 零依赖（仅用 Node 标准库）。
 * （原 tinycss2 / postcss 模式依赖外部环境，已精简；basic 是默认且为编排脚本使用的模式。）
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

/** 普通对象判定（非 null、非数组） */
function isPlainObject(value) {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * 解包外层包装结构，返回内层 DSL（页面在 page_content 内，区块在 content 内）。
 * 仅当内层确实是含 componentName 的节点时才解包；否则原样返回。
 */
function extractInnerDsl(dslData) {
  if (isPlainObject(dslData)) {
    for (const key of ['page_content', 'content']) {
      const inner = dslData[key];
      if (isPlainObject(inner) && 'componentName' in inner) {
        return inner;
      }
    }
  }
  return dslData;
}

/** CSS 语法检查器（基础模式，无需额外依赖） */
export class BasicCssChecker {
  /** @param {*} dslData */
  constructor(dslData) {
    this.dsl = dslData;
    this.errors = [];
    this.warnings = [];
  }

  /** 检查 CSS 语法 */
  check() {
    // 从外层包装(page_content/content)解包到内层 DSL 后再读取 css
    const inner = extractInnerDsl(this.dsl);
    const cssString = inner.css ?? '';

    if (!cssString) {
      this.warnings.push('No CSS field found');
      return true;
    }

    return this._checkCss(cssString);
  }

  /** 基础检查：括号匹配、基本语法 */
  _checkCss(css) {
    // 检查括号匹配
    const stack = [];
    for (let i = 0; i < css.length; i++) {
      const char = css[i];
      if (char === '{') {
        stack.push([char, i]);
      } else if (char === '}') {
        if (stack.length === 0 || stack[stack.length - 1][0] !== '{') {
          this.errors.push(`Unmatched '}' at position ${i}`);
          return false;
        }
        stack.pop();
      } else if (char === '(') {
        stack.push([char, i]);
      } else if (char === ')') {
        if (stack.length === 0 || stack[stack.length - 1][0] !== '(') {
          this.errors.push(`Unmatched ')' at position ${i}`);
          return false;
        }
        stack.pop();
      }
    }

    if (stack.length) {
      for (const [char, pos] of stack) {
        this.errors.push(`Unclosed '${char}' at position ${pos}`);
      }
      return false;
    }

    // 移除注释进行检查
    const cssNoComments = css.replace(/\/\*[\s\S]*?\*\//g, '');

    // 检查是否有 CSS 规则
    if (!cssNoComments.includes('{')) {
      this.warnings.push('CSS may not contain any rules');
    }

    // 检查分号使用
    const rules = [...cssNoComments.matchAll(/\{([^}]*)\}/g)].map((m) => m[1]);
    for (const rule of rules) {
      const properties = rule.split(';');
      // 最后一个可能为空
      for (const propRaw of properties.slice(0, -1)) {
        const prop = propRaw.trim();
        if (prop && !prop.includes(':')) {
          this.warnings.push(`Property without colon: ${prop.slice(0, 50)}`);
        }
      }
    }

    return this.errors.length === 0;
  }

  /** 生成报告 */
  report() {
    const lines = [];
    if (this.errors.length) {
      lines.push('❌ CSS Errors:');
      for (const error of this.errors) lines.push(`  - ${error}`);
    }
    if (this.warnings.length) {
      lines.push('⚠️ CSS Warnings:');
      for (const warning of this.warnings) lines.push(`  - ${warning}`);
    }
    if (this.errors.length === 0 && this.warnings.length === 0) {
      lines.push('✅ CSS check passed!');
    }
    return lines.join('\n');
  }
}

// 可用模式表（仅保留 basic）
const CHECKERS = { basic: BasicCssChecker };

function main() {
  if (process.argv.length < 3) {
    console.log('Usage: check_css.mjs <dsl-file> [mode]');
    console.log('  mode: basic (default)');
    process.exit(1);
  }

  const filePath = process.argv[2];
  const mode = process.argv[3] || 'basic';

  // 读取 DSL 文件
  let dslData;
  try {
    const text = fs.readFileSync(filePath, 'utf8');
    dslData = JSON.parse(text);
  } catch (e) {
    if (e instanceof SyntaxError) {
      console.log(`❌ Invalid JSON: ${e.message}`);
    } else if (e.code === 'ENOENT') {
      console.log(`❌ File not found: ${filePath}`);
    } else {
      throw e;
    }
    process.exit(1);
  }

  // 选择检查器
  const CheckerClass = CHECKERS[mode];
  if (!CheckerClass) {
    console.log(`❌ Unknown mode: ${mode}`);
    console.log(`Available modes: ${Object.keys(CHECKERS).join(', ')}`);
    process.exit(1);
  }

  const checker = new CheckerClass(dslData);
  const isValid = checker.check();
  console.log(checker.report());
  process.exit(isValid ? 0 : 1);
}

const __filename = fileURLToPath(import.meta.url);
if (path.resolve(process.argv[1] || '') === __filename) {
  main();
}
